scripts/versioning/check-version.ts

```diff
@@ -0,0 +1,54 @@
+// this script will be triggered from workflow actions when a new tag is push
+import {ethers} from 'hardhat'
+import {Contract} from 'ethers'
+import {ERC20Treasury} from '../../typechain'
+import {log} from '../../config/logging'
+import {deployContract} from '../common'
+
+const AddressZero = ethers.constants.AddressZero
+
+// for each deployable contract, check that the hardcoded version matches the tag version
+async function checkReleaseTag(contractName: string): Promise<boolean> {
+    const gitSourceTag = process.env.SOURCE_TAG
+    if (gitSourceTag) {
+        log.info(`git tag from workflow: ${gitSourceTag}`)
+    } else {
+        log.info(`git tag from workflow not defined!`)
+        process.exit(1)
+    }
+
+    log.info(
+        `Checking ${contractName} release against git release tag ${gitSourceTag}`
+    )
+
+    // how do we generically deploy contracts, with various constructor params
+    const contract: Contract = await deployContract<ERC20Treasury>(
```

Comment:
question: contracts can either be deployed directly or with a proxy.

The existing scripts I'd threw together to deploy directly to Rinkbey, but the actual prod script will be a combination of direct and proxy deployments.

Reply:
We could make a convention to have a script for every contract (as each could have bespoke args) 🤔

Reply:
I've refactored so that the function takes an already constructed `Version` contract type, calls `VERSION()` on this contract and checks against the git tag - seems like a better separation of concerns to have the contract already constructed. 

Not sure how this fits with proxy deployments - are we talking Proxy Factories or just pure Proxies? Will the new approach ive taken be able to also just call `VERSION()` on the proxy - same process?

Have used the `hardhat-deploy` plugin for deployment with mixed results in the past. (https://github.com/wighawag/hardhat-deploy)

Reply:
>Not sure how this fits with proxy deployments - are we talking Proxy Factories or just pure Proxies? Will the new approach ive taken be able to also just call VERSION() on the proxy - same process?

Both need to be supported, as this will become out generalized approach.

>Have used the hardhat-deploy plugin for deployment with mixed results in the past. 

What's issues have you encountered previously with HardHat deploy?

I'm sceptical of a few of the claims on (https://github.com/wighawag/hardhat-deploy), in particular deterministic deployment across networks (as you can't guarantee that, i.e. the premise being you have the same state across networks, which does always hold true, and when deviance occurs you don't have deterministic deploy )

With JS/TS experience has taught me to run lean on dependencies, as whatever level of support there is today can vanish overnight.

Reply:
> Both need to be supported, as this will become out generalized approach.
Should this ultimately go in the solidity template repo, or optional there?

> What's issues have you encountered previously with HardHat deploy?
- the gas price estimation can be borfed and generally requires overwriting/hardcoding the gasPrice when the network is busy. general criticism i guess - this can be tricky generally.
- very little testing in the repo & tries to do too much.

What was nice about it was having a `deploy` & `deployments` directory. Easier to make deployments idempotent, with the recentlyDeployed flag.
Never used that deterministic deployment feature (create2?). 

I like the ethers contract factory and less dependences. Definitely not a hill i would die on lol.